### DISCUSSION TITLE
Allow singleline expressions in try-catch blocks

### DISCUSSION THREAD

**Discussion by weitzhandler** [dotnet/csharplang#4676] [Language Ideas]
Please allow omission of curly braces for single-statement try-catch blocks. I'm sure this small syntactic sugar has been proposed before, although I couldn't find it.

    try
      client.Load();
    catch (Exception e)
      logger.Log(e);
    finally
      client.Dispose();

Just like:

    if(!isValid)
      ShowError("error");

**Comment by yaakov-h** [comment]
@lachbaer Did you ever create the two separate proposals to follow on from #220?

**Comment by lachbaer** [comment]
@yaakov-h No, I thought that it wouldn't be worthwhile, because I have inspected some code and couldn't see how `try` expressions can be really useful. Most times a refactoring and more thorough exception handling seemed to be the better solution. Catch expression syntax `catch (...) => return;` however could be useful. I just haven't opened a topic for it.

And `catch` expressions clash with `try` expressions anyway, because they could be nested and then it will lead to `catch` ambiguity again, the reason why `try`-`catch` allow block statements only in the first place.

**Comment by weitzhandler** [comment]
@yaakov-h, I actually saw @lachbaer's proposal, but since it was closed and my proposal is a mere syntax sugar I thought about posting it separately having not found another one.

**Comment by yaakov-h** [comment]
A big issue we had in #220 was if you nest them, you have ambiguous behaviour.

For example:

```
try
    try
        DoTheThing();
    catch (ThingException e)
        HandleTheThingFailing();
  catch (Exception e)
      // What scope am I in now? This is ambiguous.
  finally
      // How about now?
    try
        // So this would be second-level for sure, right?
    catch (Exception e)
        // As would this
  finally
  // This would belong to the outer block if the first finally belongs to the inner one, otherwise the entire second try/catch/finally is not nested at all.
```

**Comment by weitzhandler** [comment]
1) The same problem is in `if`-`else` and `else if`.
2) It can be restricted to inline expressions, like:
    public string MyProperty => try { } catch { } //invalid

**Comment by lachbaer** [comment]
Of course rules that define the level of nesting could be defined, such as a `finally` always breaking the current level, giving control back to the previous `try`, `catch` or `finally` statement.

But because `try`-`catch` shall be a construct that clearly structures (mainly unexpected) error cases every further nesting only makes the flow more unclear.

If you only have one line that you want to `try` over, then you can just put the `try`-block in the same line, as long as it doesn't violate your style rules.
```cs
public void Method()
{
    int value;
    try { value = ErrorproneFunction(); }
    catch
    {
        // structured exception handling block
    }
}
```
This also *emphasizes* the explicit `try`.

Another scenario is a one-line block-less `catch`, because this seems to be more common. It would only reduce the occurance of 2 curly braces, but looks a bit better.
```cs
    try {
        statement1;
        statement2;
        statementN;
    }
    catch (IOException ex) => throw new MyIOException(ex);
    catch (NullReferenceException ex) => return;
```
Because those are *expressions* a nesting cannot occur (as long as try-expressions aren't ready).

**Comment by DavidArno** [comment]
> The same problem is in `if-else` and `else if`.

Is that not a good reason to avoid adding those same problems to `try catch`?

**Comment by weitzhandler** [comment]
I agree but I don't see it as a problem. It's habit of the way we're predict our code. And since this habit already exists in the language and we're used to deal with it, I see no problem in having the same in other areas. Nested `try`s should be in braces just like nested `if`s. Besides, there is option 2, of avoiding that by allowing only expressions.

**Comment by HaloFour** [comment]
@weitzhandler 

The blocks are required because, unlike `if`/`else`, a `try` block can have any number of `catch` blocks.  If you were to nest a single-line `try` under another single-line `try` and have more than two `catch` blocks there'd be no way for the compiler to determine which `try` statements that each of the `catch` blocks are associated with.

**Comment by weitzhandler** [comment]
So what?
Either a block or a single statement (or an empty one - e.g. `{ }`) is required in any `catch` statement, just like it is now.

**Comment by HaloFour** [comment]
`try try Foo(); catch (IOException) Bar(); catch (Exception) Baz(); catch Console.WriteLine("oops");`

What's it do?  There's two perfectly legal answers.

**Update**:  Missing some necessary semicolons, fixed.

**Comment by YaakovDavis** [comment]
> ```
> try
>   client.Load();
> catch (Exception e)
>   logger.Log(e);
> ```
> Just like:
> ```
>  if(!isValid)
>    ShowError("error");
> ```

The `if` statement ends with the `;`, whereas the `;` after the `client.Load()` call doesn't end the `try` statement as a whole. 
This makes the cases somewhat distinct from a visual perspective.

**Comment by weitzhandler** [comment]
@HaloFour 
> `try try Foo();`

nesting `try` should only be allowed in blocks, has been discussed above
<hr/>

@YaakovDavis 
>The `if` statement ends with the `;` whereas the `;` ... doesn't end the `try` statement as a whole.

It ends the `catch` statement as a whole.
And the `if` doesn't completely end, `else if` statements can be following it.

**Comment by HaloFour** [comment]
@weitzhandler 

> nesting `try` should only be allowed in blocks, has been discussed above

So you'll need two flavors of the C# spec to deal with this possibility?  Afterall, there could be any number of other nestable constructs under that first `try`.  Expressing that in the spec would be obnoxious.  Not to mention it would make refactoring a pain since the legality of the syntax would depend entirely on code external to that statement.

It's simply not worth it.  If you want it all on one line feel free to wrap it in curly braces on that one line.

**Comment by lachbaer** [comment]
> nesting try should only be allowed in blocks

In a statement's place *every* 'statement' is allowed, so `try` is allowed, too. Putting a restriction on this rule *only* on `try`, `catch` and `finally` would be quite an incision.

**Comment by YaakovDavis** [comment]
> And the `if` doesn't completely end, `else if` statements can be following it.

Good point, I forgot about this (trivial) case.

**Comment by HaloFour** [comment]
`else if` is two statements, the grammar lends itself to that formatting short cut but that's not how it's parsed or interpreted.

**Comment by lachbaer** [comment]
> else if is two statements,

One (1) `else` **always** ends the previous `if-else` construct! `catch` however does not, further `catch`es can follow.

But after a `finally` the `try` construct definitely ends, so a `finally`-block *could* be treated as a break to the previous level.

**Comment by weitzhandler** [comment]
@HaloFour 
> So you'll need two flavors of the C# spec to deal with this possibility? 

@lachbaer 
>In a statement's place every 'statement' is allowed

This was discussed above. There is the flavor of `void DoAction() => WriteLine("");` where you cannot go `void DoAction() => try`, try-blocks don't count as a single-line statement. So I'm not inventing any new flavors.

@lachbaer 
>One (1) else always ends... But after a finally...

That is correct but `if` doesn't require an `else`, just like `try` doesn't require a `finally`, so I see multiple `catch`es as multiple `else if`s.

**Comment by lachbaer** [comment]
I'm 'catch'ing up @HaloFour 's example from above and add an empty `finally` statement.
```cs
try try Foo(); catch (IOException) Bar(); catch (Exception) Baz(); finally ; catch Console.WriteLine("oops");
```
This would have one definite outcome, if only `finally` can break the nesting. But even if this *was* allowed, **don't do that** :-)

**Comment by lachbaer** [comment]
>  There is the flavor of void DoAction() => WriteLine(""); where you cannot go void DoAction() => try

After the *arrow token* `=>` only *expressions* are allowed, `try` is a **statement**.

Try defining this method in C# 7:
```cs
public void Test() => while(true);  // Error CS1525 Invalid expression term 'while'
```
, because `while` is a statement - like `try` - and not a statement.

**Comment by lachbaer** [comment]
> [selfquote:] After the arrow token => only expressions are allowed, try is a statement.

Besides, that's why I proposed `catch => expression;` above. It cannot introduce another nested `try` block.

Addendum: adding `try => expressions;` counters the proposal of `catch => expression;`, because you will end up with with the same dilemma.
```cs
try => try => try => try => expression; catch => try => expression; finally => expression;
    catch => try ...
```

**Comment by weitzhandler** [comment]
@lachbaer [the other comment](https://github.com/dotnet/csharplang/issues/616#issuecomment-303106862)
> I'm 'catch'ing up @HaloFour 's example

You're right, this is the same issue as we have with nested `if`s, so we could either limit to _expressions_, or have the same behavior as with nested `if`s which I have no problem with, and I rather prefer one-liners and less braces.
Those with more braces coding style guidelines will always have the option to include braces.

> [selfquote:] After the arrow token => only expressions are allowed, try is a statement.

I personally prefer without the `=>`, but I hear your point loud and clear.

**Comment by lachbaer** [comment]
> .. or have the same behavior as with nested if

No, you cannot! Imagine every `else` was an implicit `else if ` (purely figurative) then you can end up with,
```
if (cond)
    statement;
else
    statement;
else
    statement;
else
    if (cond)
        statement;
    else
        statement;
else
    statement;
```
If there weren't the indentions that would be really hard to read and understand! And a `try` *can* have multiple `catch`.

**Comment by weitzhandler** [comment]
Your indentations are valid. So the `try`-`catch`-`finally` issue is similar to the `if`-`else-if`-`else` statement, isn't it?

**Comment by lachbaer** [comment]
Similar, not equal.

**Comment by weitzhandler** [comment]
I get the readability issue tho I don't understand the difference, a nested `if`-`else if` will always need to be wrapped in braces in order to escape to allow an `else`/`else if` of parent, same with nested `catch`es.

**Comment by bondsbw** [comment]
```c#
if (...)
if (...) Foo();
else Bar();
else Baz();
else Console.WriteLine("oops");
```

That isn't valid C#.  `else Bar();` goes with the second `if`.  `else Baz();` goes with the first `if`.  `else Console.WriteLine("oops");` does not have an `if`.

There seems to be confusion about `else if` but keep in mind that there is no such thing as `else if`.  It is an `else` whose statement happens to be an `if`.  
```c#
else if (b) { ... }
```
 is just removing some curly braces in
```c#
else { if (b) { ... } }
```

**Comment by shravan2x** [comment]
I find myself writing code like this often:

```
try
{
    return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path));
}
catch
{
    return new ArgonSettings();
}
```

This occupies 4 more lines than I would like it to, using up more space and contributing no value. It would be nice if this feature was added to C#.

**Comment by HaloFour** [comment]
@shravan2x 

> This occupies 4 more lines than I would like it to,

So why not write it like this?

```cs
try { return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path)); }
catch
{
    return new ArgonSettings();
}
```

**Comment by shravan2x** [comment]
@HaloFour It seems asymmetrical and contradicts its own coding style. Why is one pushed to a single line while the other is not? Both of them are the same format `<keyword> { return <something>; }`.

Next, I want to avoid braces that aren't necessary. They add an extra level of indirection when reading the code, which is not ideal.

**Comment by CyrusNajmabadi** [comment]
So why not have: 

```c#
try { return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path)); }
catch { return new ArgonSettings(); }
```

?

**Comment by CyrusNajmabadi** [comment]
I think the point of the question is: if you don't want these taking up multiple lines, the solution is simple: don't spread the existing construct over multiple lines.

**Comment by shravan2x** [comment]
@CyrusNajmabadi

> So why not have

Because the extra braces are unnecessary, and they make reading difficult. I'd like to avoid boilerplate that serves no purpose.

> I think the point of the question is: if you don't want these taking up multiple lines, the solution is simple: don't spread the existing construct over multiple lines.

If I understand correctly, you're suggesting that I _force_ single lines. This makes the code _harder_ to understand, not _less_. Further, these don't agree with my existing code style, which avoids braces for all one-line blocks.

At the end of the day, I suspect it will get harder to reason about, since it will fall to personal preference. A bigger question of mine is: _Why does try-catch lack this functionality when if-else, while, do-while, for, for-each and others have had it forever?_

**Comment by HaloFour** [comment]
@shravan2x

> Because the extra braces are unnecessary, and they make reading difficult. I'd like to avoid boilerplate that serves no purpose.

The braces are necessary to keep consistency in parsing with the `catch` block where braces are required to handle the nested `try` problem.  Two characters is not an onerous amount of "boilerplate".

> This makes the code harder to understand, not less. Further, these don't agree with my existing code style, which avoids braces for all one-line blocks.

It doesn't make the code and more difficult to read than omitting the braces would.  Why should the language spec and compiler be modified to suit your inflexible stylistic preferences?  Code styles aren't law; they're only guidelines to aid in code legibility.

**Comment by CyrusNajmabadi** [comment]
@shravan2x What do you want it to look like?  You did not mention in your original post.

**Comment by shravan2x** [comment]
@HaloFour

>  The braces are necessary to keep consistency in parsing with the catch block where braces are required to handle the nested try problem. Two characters is not an onerous amount of "boilerplate".

Not necessarily; we could just let the same thing happen as with nested if-else statements. If someone wants to use nested, single-line try-catches, they have to use braces. That's all. There's no reason to say no to the entire feature for one single edge case. 

Two characters are not much, I agree with you there. But it's not the file size I'm worried about - it's the clutter in my mind. Think of it as pushing on a mental stack every time you enter a nested block. It gets tedious very fast.

> It doesn't make the code and more difficult to read than omitting the braces would.

As I said, that is _your_ opinion. In _my_ (and clearly, many, many others) minds, it _does_ make it more difficult.

>  Why should the language spec and compiler be modified to suit your inflexible stylistic preferences?

Again, this is not _only my preference_. The mere existence of this, and all duplicate threads, suggests that other people share this view. In fact, it would seem that _you're the one who is being inflexible_, since this feature will not break any existing code, and anyone who wishes to use the existing syntax freely can.

> Code styles aren't law; they're only guidelines to aid in code legibility.

I agree with this 100%. So why not do what it takes to make code more legible, and get rid of the braces?

@CyrusNajmabadi 

> What do you want it to look like? You did not mention in your original post.

This is what I had in mind, but of course, there may be clearer ways of doing it:
```
try
    return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path));
catch
    return new ArgonSettings();
```

**Comment by HaloFour** [comment]
@shravan2x 

> This is what I had in mind, but of course, there may be clearer ways of doing it:

It's already been declared, and well documented since C# 1.x, that the syntax that you have suggested is inherently ambiguous.  If the `try` statement is nested it is completely impossible for the compiler to correlate `catch` statements to the correct `try` statement.  It cannot be compared to `if`/`else` because a single `if` statement can only have at most a single `else` statement, but a `try` statement can have any number of `catch` statements.  It's not possible to remove the requirement of braces for the `catch` statement.

**Comment by shravan2x** [comment]
@HaloFour 

As I mentioned earlier, you seem to be disregarding the entire idea for one single edge case. What if we just allowed the first following catch to use the blockless format? i.e. if you had nested try-catches with multiple catch statements, only allow the first one (and maybe display a warning). This is just one way to handle it. There may be other better ones, but irrespective, it needn't hold back the entire feature.

**Comment by HaloFour** [comment]
@shravan2x 

Nested exception handling is pretty common, not an edge case.  You're suggesting that the language should become internally inconsistent with itself just to satisfy your stylistic preferences.  That's not remotely close to satisfying the exceptionally high bar required to modify the language, compiler and all supporting tooling.

**Comment by gafter** [comment]
Just to be clear, I understand this proposal as asking for the introduction of a new ambiguity into the C# language for the purpose of occasionally saving four characters in programs.

Is the proposal really so much better than

    try
      { client.Load(); }
    catch (Exception e)
      { logger.Log(e); }

**Comment by shravan2x** [comment]
@gafter It indeed is.
@HaloFour Again, it clearly not _only my_ stylistic preference. It doesn't become any more inconsistent than nested single-line if-else statements already work. Do you propose removing those too?

**Comment by jnm2** [comment]
@shravan2x There's a big difference between deciding not to add a new feature that will cause confusion and deciding to make people's existing code break.

**Comment by shravan2x** [comment]
@jnm2 How does it cause any more confusion than the existing nested single-line if-else statements?

**Comment by HaloFour** [comment]
@shravan2x

> How does it cause any more confusion than the existing nested single-line if-else statements?

Because `try` can have any number of `catch` clauses whereas `if` can only ever have zero or one `else` clauses.  It's not only confusing to read to a developer, it's also impossible to parse unambiguously to the compiler.

But yes, nested single-line `if` statements can cause confusion, especially when you don't have an IDE that will try to correct your nesting indentation.  It's legacy syntax inherited from C and can't be removed from the language.  C++ learned from that mistake when they added `try`/`catch`.

**Comment by jnm2** [comment]
@shravan2x How do two things, one of which starts appearing in places people aren't expecting to see it, cause more confusion than one thing?

**Comment by shravan2x** [comment]
@jnm2 They're the same thing.

```
if condition
    if condition2
        statement1;
    else
        statement2;
```

@HaloFour 
> Because try can have any number of catch clauses whereas if can only ever have zero or one else clauses. It's not only confusing to read to a developer, it's also impossible to parse unambiguously to the compiler.

I agree that it's confusing, but _it has already been added_. In an ideal world, I wouldn't have seen them either. Including single-line try-catch statements, however, improves style compatibility with existing code that uses single-line if-else statements.

If anything, we should be discouraging users from using single-line versions of these statements, not denying them access to it (Similar to using pointers in C# or such).

**Comment by HaloFour** [comment]
@shravan2x 

> I agree that it's confusing, but it has already been added.

Not for `try`/`catch` it hasn't.  And, as I've repeated several times, for good reason.  It's impossible for the compiler to unambiguously parse it.  For that reason alone it *can't* be added because there's nothing that the compiler can do.

The following may be considered bad practice, but at least it can only have exactly one meaning:

```cs
if condition
    if condition2
        statement1;
    else
        statement2;
```

But the following can mean literally *three* semantically different but perfectly legal things:

```cs
try
    try DoSomething()
    catch (FooException e) HandleFoo(e);
    catch (BarException e) HandleBar(e);
    catch (BazException e) HandleBaz(e);
finally CleanUp();
```

There's simply no correct way for the C# compiler to parse that.  The only `catch` clause that can be unambiguously associated with a `try` clause is the first one.  All of the rest could belong to either `try` clause.

And that aside, the fact that `if`/`else` does support embedded statements that aren't blocks is not reason to repeat that mistake elsewhere.

**Comment by HaloFour** [comment]
@shravan2x 

Bjarne Soustroup [offered the same explanation](http://www.stroustrup.com/except89.pdf) as to why the braces are also required for `try`/`catch` clauses in C++:

> Allowing exception handlers to be attached to blocks only and not to
simple statements simplifies syntax analysis (both for humans and
computers) where several exceptions are caught and where nested
exception  handlers are considered (see Appendix E). For example,
assuming that we  allowed handlers to be attached to any statement
we could write:
>
> `try try f(); catch (x) { ... } catch (y) { ... } catch (z) { ... }`
>
> The could be interpreted be in at least three ways:
>
> ```
> try { try f(); catch (x) { ... } } catch (y) { ... } catch (z) { ... }
> try { try f(); catch (x) { ... } catch (y) { ... } } catch (z) { ... }
> try { try f(); catch (x) { ... } catch (y) { ... } catch (z) { ... } }
> ```
>
> There seems to be no reason to allow these ambiguities even if there
> is a trivial and systematic way for a parser to chose one
> interpretation over another. Consequently, a `{` is required after a
> try and a matching `}` before the first of the associated sequence of
> catch clauses.

**Comment by CyrusNajmabadi** [comment]
> This is what I had in mind, but of course, there may be clearer ways of doing it:

```c#
try
    return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path));
catch
    return new ArgonSettings();
```

I'm not seeing htat s substantively shorter or better than either:

```c#
try
{ return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path)); }
catch
{ return new ArgonSettings(); }
```

Or

```c#
try {
    return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path));
} catch {
    return new ArgonSettings();
}
```

If you want terse code, you can get it.  

Note: C# does not generally remove existing syntactic tokens just because they can be removed.  We will do it when there is a fairly substantial saving.  For example, Properties used to have to be written as:

```c#
int _prop;
int Prop { get { return _prop; } set { _prop = value; } }
```

But we can now write that as:

```c#
int Prop { get; set; }
```

In this case we've saved 50 chars over usage of a *very* common construct.  We've also distilled the important parts of the code out, making things clearer.  This is a case of a clear worthwhile win.

I'm just not seeing it with try/catch.  We're talking about a couple of characters saved.  And for a construct that is already spread over several lines, and which already involves numerous statements and clauses.  

It's like asking to make type parameter brackets optional in a class/struct/interface declaration.  i.e. ```public class Dictionary TKey, TValue { }```.  Sure, it could be done.  But is it really helping, and is the syntactic brevity *substantially* benefitting the code.  To me, that's just not happening here.

**Comment by Thaina** [comment]
I just got desire to have this feature for remove indentation

I mean if we support this feature we could write

```C#
try
    foreach(var item in list)
    {
        DoSomeThing(item);
    }
catch(Exception e)
{
}
```

Which is maybe up to formatter but it could transform into

```C#
try foreach(var item in list)
{
    DoSomeThing(item);
}
catch(Exception e)
{
}
```

**Comment by mkane91301** [comment]
I would think that if:
```C#
try => DoSomething();
catch (Exception ex) => HandleError(ex);
```
would be equivalent to:
```C#
try { DoSomething(); }
catch (Exception ex) { HandleError(ex); }
```
that would handle any ambiguity and would solve the aesthetic problem of the curly brace noise.

And it would be very nice in the case where DoSomething returns a value, if you could write:
```C#
var result = try => DoSomething();
             catch (Exception ex) => HandleError(ex);
```

**Comment by yaakov-h** [comment]
@mkane91301 how does that solve ambiguity with nested `try` expressions?

**Comment by shravan2x** [comment]
@yaakov-h 
> how does that solve ambiguity with nested try expressions?

It doesn't need to. Just like with single-line if-else statements, we pick one disambiguation strategy and clearly state it in the docs. The fact that some potential edge cases exist is not a good reason to discard the entire idea.

@CyrusNajmabadi 
> I'm not seeing htat s substantively shorter or better than either:

I (and clearly many others) disagree. This code:
```c#
try
{ return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path)); }
catch
{ return new ArgonSettings(); }
```

is significantly more difficult to read than:
```c#
try
    return JsonConvert.DeserializeObject<ArgonSettings>(File.ReadAllText(path));
catch
    return new ArgonSettings();
```

The objective is to have code that's easier to understand at first glance. Simply compressing existing code into fewer lines doesn't solve that problem. And for the projects that prefer the former, they could simply pick that as a setting in code style preferences.

**Comment by HaloFour** [comment]
@shravan2x 

> It doesn't need to. Just like with single-line if-else statements, we pick one disambiguation strategy and clearly state it in the docs. The fact that some potential edge cases exist is not a good reason to discard the entire idea.

The disambiguation strategy was already chosen.  The strategy is to not introduce ambiguity in the first place.  It's not worth it to save 4 characters.

**Comment by shravan2x** [comment]
@HaloFour Maybe that's a good idea, maybe it isn't. But that decision should have been made _before_ adding support for it with if-else statements. At this point, code (and the language as a whole) looks inconsistent when different rules have to be followed for different blocks of code.

In addition, the ambiguity itself only exists in edge cases. Like I mentioned before, the fact that some potential edge cases exist is not a good reason to discard the entire idea.

> It's not worth it to save 4 characters.

Clarity of code is not measured solely by the number of characters.

**Comment by theunrepentantgeek** [comment]
>  But that decision should have been made before adding support for it with if-else statements. 

The way that if-else statements works was carried forward from C (in fact, possibly from it's predecessor) so it's literally been around 50 years. Effectively this is grandfathered in to peoples thinking. 

> Clarity of code is not measured solely by the number of characters.

Yep. Totally agree with this. 

Which is why I don't believe this idea has any merit - the ***only*** benefit it has is to reduce the number of characters typed, with an increased cognitive load for everyone reading the code.

Clarify of code is _impaired_ by having ambiguous statement structures.

Regrettably it can't happen because it would be a major breaking change, but I'd prefer to go the route chosen by Go (and other languages) where the `{` braces `}` are _required_ everywhere.

**Comment by shravan2x** [comment]
> Effectively this is grandfathered in to peoples thinking.

I agree, which is why it feels unintuitive when people use the same logic to try-catch statements and realize it doesn't work.

> the only benefit it has is to reduce the number of characters typed, with an increased cognitive load for everyone reading the code.

I don't agree that the only benefit is to reduce the number of characters typed. Having fewer characters is certainly a side-effect, but the main reason is that removing braces **reduces** the cognitive load for people reading the code.

I understand you disagree with this, and people may well have divided opinions on it. But there certainly are a significant portion of people on this side of that divide. Individual teams could make this choice for themselves by setting code style preferences.

> Clarify of code is impaired by having ambiguous statement structures.

Not all forms of this syntax are ambiguous. There are many valid use-cases for this feature that work well.

It _can potentially_ be ambiguous when used in certain ways (which is also the case with many other C# features), the responsibility is on the developer to make the right choice in each situation.

> Regrettably it can't happen because it would be a major breaking change, but I'd prefer to go the route chosen by Go (and other languages) where the { braces } are required everywhere.

That's an interesting opinion, I don't entirely disagree with it.

**Comment by HaloFour** [comment]
@sachevl 

> but the main reason is that removing braces **reduces** the cognitive load for people reading the code.

Punctuation matters.  Removing it **increases** the cognitive load because now the reader has to understand how the compiler is going to parse those statements and whether or not the whitespace used by the author actually matches those rules.  Relying on trivial whitespace is how bugs get introduced.  The braces serve as punctuation and forces disambiguation _for the reader_.

This was already clear 30 years ago, which is why C++ has never had a form of `try` which allowed for the omission of braces.

**Comment by CyrusNajmabadi** [comment]
> But that decision should have been made before adding support for it with if-else statements. 

If we had a time machine, perhaps someone could go back and work to make that happen 23 years ago.  :)  

> looks inconsistent when different rules have to be followed for different blocks of code.

As someone who knows the C# grammar inside and out, i can tell you *it is massively inconsistent*.  To the point where i would practically argue that consistency is a non-goal (or very minor goal).  When we design new stuff we consider this a bit, but it's not what we're optimizing for.  Furthermore, for stuff that has already shipped, we're not going to change it just for the sake of consistency.  There has to be serious value delivered on top of that for it to happen.  I'm just not seeing it here.

**Comment by shravan2x** [comment]
@HaloFour 
> Punctuation matters. Removing it increases the cognitive load because now the reader has to understand how the compiler is going to parse those statements and whether or not the whitespace used by the author actually matches those rules.

Removing braces **increases** the cognitive load because the reader now has to create a new mental stack frame for each new block in the try-catch statement. If there's just a single statement, it's simpler to read because it doesn't create a new frame. The feature is sufficiently disambiguous, and clearer for the reader.

But like I said before, people may well have divided opinions on this. Your opinion need not necessarily be held by every user of C#. Clarity is a subjective measure and code isn't clearer to _everyone_ simply because it's clearer _for you_. There are a significant number of users who find it clearer without the braces. Individual teams could make this choice for themselves by setting code style preferences.

@CyrusNajmabadi 
> If we had a time machine, perhaps someone could go back and work to make that happen 23 years ago. :)

If only! :)

> As someone who knows the C# grammar inside and out, i can tell you it is massively inconsistent. To the point where i would practically argue that consistency is a non-goal (or very minor goal). When we design new stuff we consider this a bit, but it's not what we're optimizing for.

You may be right, but my opinion is that working to smooth out kinks is always a good direction to aim for, even if it takes a lower priority. I certainly don't expect that this feature should outweigh others that the community finds better value in.

> Furthermore, for stuff that has already shipped, we're not going to change it just for the sake of consistency. There has to be serious value delivered on top of that for it to happen. I'm just not seeing it here.

That's a fair point - this feature is functionally just syntactic sugar. But consistency is not the only reason, clarity is another important one. A number of new features (like Expression Bodied Members from C# 6) fall in similar categories, but were still added.

**Comment by CyrusNajmabadi** [comment]
> A number of new features (like Expression Bodied Members from C# 6) fall in similar categories, but were still added.

This falls into the "There has to be serious value delivered" bucket.  There's likely a 1000x (or more) benefit to expression bodied members versus braceless-try-statements.

**Comment by shravan2x** [comment]
> This falls into the "There has to be serious value delivered" bucket. There's likely a 1000x (or more) benefit to expression bodied members versus braceless-try-statements.

Why do you hold this opinion? Doesn't it also just avoid two extra braces and a return statement i.e. creating a block of code?

**Comment by CyrusNajmabadi** [comment]
> Why do you hold this opinion?

Because the amount of code that uses properties (esp. very trivial ones) is massively more than the amount of code that needs to use trivial try/catches.

**Comment by shravan2x** [comment]
@CyrusNajmabadi Fair point. Another example of a feature that arguably delivers less value than this one is [Enhancement of interpolated verbatim strings](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#enhancement-of-interpolated-verbatim-strings).

**Comment by 333fred** [comment]
> @CyrusNajmabadi Fair point. Another example of a feature that arguably delivers less value than this one is [Enhancement of interpolated verbatim strings](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#enhancement-of-interpolated-verbatim-strings).

And that issue has significantly less cost and doesn't result in complicated and hard-to-reason-about ambiguity resolution rules.

**Comment by Thaina** [comment]
> @shravan2x
> 
> > I agree that it's confusing, but it has already been added.
> 
> Not for `try`/`catch` it hasn't. And, as I've repeated several times, for good reason. It's impossible for the compiler to unambiguously parse it. For that reason alone it _can't_ be added because there's nothing that the compiler can do.
> 
> The following may be considered bad practice, but at least it can only have exactly one meaning:
> 
> ```cs
> if condition
>     if condition2
>         statement1;
>     else
>         statement2;
> ```
> 
> But the following can mean literally _three_ semantically different but perfectly legal things:
> 
> ```cs
> try
>     try DoSomething()
>     catch (FooException e) HandleFoo(e);
>     catch (BarException e) HandleBar(e);
>     catch (BazException e) HandleBaz(e);
> finally CleanUp();
> ```
> 
> There's simply no correct way for the C# compiler to parse that. The only `catch` clause that can be unambiguously associated with a `try` clause is the first one. All of the rest could belong to either `try` clause.
> 
> And that aside, the fact that `if`/`else` does support embedded statements that aren't blocks is not reason to repeat that mistake elsewhere.

What if we strictly define that the try-catch block muct end with `;` ?

```
try
    try DoSomething()
    catch (FooException e) HandleFoo(e)
    catch (BarException e) HandleBar(e)
    catch (BazException e) HandleBaz(e); // inner
finally CleanUp(); // outer

// comparison

try
    try DoSomething()
    catch (FooException e) HandleFoo(e)
    catch (BarException e) HandleBar(e); // inner
catch (BazException e) HandleBaz(e)
finally CleanUp(); // outer

// one line

try
    try DoSomething() catch (FooException e) HandleFoo(e); // inner
catch (BarException e) HandleBar(e)
catch (BazException e) HandleBaz(e)
finally CleanUp(); // outer
```

Or we could also use `,` like a switch expression to separate between catch expression?

**Comment by cnapolit** [comment]
Is there any reason why requiring curly braces for nested try-catch statements that are not already contained in a code block wouldn't solve the ambiguity?
```C#
try
{
    try
       action();
   catch (Exception e)
       Cleanup(e);
}
catch (Exception e)
    Cleanup(e);
try
{
    foreach(var variable in variables)
        try
            action(variable);
        catch (Exception e)
           Cleanup(e);
}
catch (Exception e)
    Cleanup(e);
```
Otherwise, any nested try inside of a nested block would not be ambiguous by the logic 'this try is linked to all immediate catches within the given code block':
```C#
// Admittedly, 'try foreach' kind of reads like:
// "upon each iteration, try the given code block"
// Rather than:
// "try this foreach statement"
// which is not good for new C# users
try foreach(var variable in variables)
{
    try
        action(variable);
    catch (Exception e)
       Cleanup(e);
}
catch (Exception e)
    Cleanup(e);
```
I would wager the biggest con just from a syntax point of view is that users unfamiliar with this form of try-catch could be confused both what it is doing and why their compiler/IDE is forcing them to use curly braces when they first attempt to use it.

If this took little to no effort to implement and analyze, I would be in favor of it.

However, given:
- Just about any change to a language is non-trivial
- Analysis is more complex than just 'find the previous if in the current block'
- The subjective syntax benefit
- The controversy around the subject

it's unsurprisingly not a good idea to implement try statements (and thus catch statements, for consistency) without curly braces.

**Comment by notato8** [comment]
The syntax may be subjective, but the proposal is not. It's backwards compatible, so each developer would either benefit from the addition, or experience no change at all. This is an objective improvement.

I agree that countless examples of unreadable code have been given, which would be enabled by this proposal. This is negligible when you consider that many developers would use the proposal to enhance the readability of their code. Analyzers exist to solve this problem.

The precedent of nested `if` statements is used an example to argue that this proposal could create ambiguity with `catch`  blocks. However, this problem has been clearly solved in the specification, which states

> An else part is associated with the lexically nearest preceding if that is allowed by the syntax.

Applying the same principle to `try-catch` statements leads to the conclusion: any `catch` block is associated the nearest preceding `try` block that hasn't been concluded by a `finally` block. There was never any ambiguity to begin with.

**Reply by CyrusNajmabadi** [reply]
@notato8dx see https://github.com/dotnet/csharplang/discussions/4676#discussioncomment-635609

**Reply by CyrusNajmabadi** [reply]
> This is an objective improvement.

It is not, as brevity is not objectively better.

**Reply by notato8** [reply]
@CyrusNajmabadi 

> > This is an objective improvement.
> 
> It is not, as brevity is not objectively better.

My stance is not that brevity is objectively better. It's that giving developers the option is objectively good.

> @notato8dx see [#4676 (comment)](https://github.com/dotnet/csharplang/discussions/4676#discussioncomment-635609)

It's not just about saving a few characters from removing the brackets. I actually agree that the one-line examples you offered are less readable. The benefit would come from reducing the complexity of nesting multiple block statements. Consider these common examples:

```
if (condition) {
    ...
} else if (condition) {
    ...
} else {
    ...
}
```
```
using (disposable)
using (disposable) {
    ...
}
```

In the opinions of many developers (maybe not to the same extent with the `using` statement example), these are more readable than the alternative of strictly using blocks:

```
if (condition) {
    ....
} else {
    if (condition) {
        ...
    } else {
        ...
    }
}
```
```
using (disposable) {
    using (disposable) {
        ...
    }
}
```

With `try`, `catch`, and `finally` blocks, this would enable things like:

```
try {
    ...
} catch (Exception) {
    ...
} finally try {
    ...
} catch (Exception) {
    ...
}
```

```
try {
    ...
} finally if (condition) {
    ...
}
```

Also, this has the side bonus that you can have an empty catch statement without brackets.

```
try {
    ...
} catch (Exception);
```

**Reply by CyrusNajmabadi** [reply]
> It's that giving developers the option is objectively good.

That is not our stance. As I mentioned in my linked post, there are a huge number of places we could make tokens optional. Even though I'm an those cases it would be developer choice, we still would view it as actively bad for the language and ecosystem to have all the confusing forms. 

We are judicious where we allow flexibility.

**Reply by CyrusNajmabadi** [reply]
> With try, catch, and finally blocks, this would enable things like:

Tbh, all those cases make me *less* in support of this proposal.

**Comment by Thaina** [comment]
I think single line try better be considered as a subset of try expression https://github.com/dotnet/csharplang/discussions/2734

Suppport try expression that can set variable or return value would be a lot more useful
